### DISCUSSION TITLE
Static Flavors / Static Dark-Light Theme support for dynamic generated css

### DISCUSSION THREAD

**Discussion by BierDav** [vercel/next.js#36502] [Ideas]
### Describe the feature you'd like to request

Next.js is currently nearly perfect when it comes to ssg, but there is one big feature missing for all emotion users (maybe other similar packages too). You really can't add a dark-light mode toggle to your website because it breaks everything. The problem is that you really can just make one statically generated version of a component/document/app for light or dark theme. If you are just using emotion/css to achive your dark and light theme, you may send two emotion caches to the client and it should decide which to use. This is not optimal in every sense.

### Describe the solution you'd like

I've two idea how to solve this problem:

Allow generating different flavors of an component/document/app with different props. Then decide in the on getServerSideProps which flavor to use. Depending on the request data (Cookie, JWT-Token, other HTTP-Header, .. ) you can directly serve the dark mode version for example.

Provide a way to be able to mess with the initialprops for each request. So that you can save both flavors into the intialprops with prerendering. And then decide which initial props to use.

I know they are both kinda similar, but the second one opens more possiblities. The first one is the more safe way

### Describe alternatives you've considered

If there is a work around I haven't found yet, please tell me. We really need this. We want to add an dark mode to our homepage, but can't without disabling ssg and only use ssr.

**Comment by GabenGar** [comment]
The basic problem with SSR and themes is if you want to avoid [the flashes of unstyled content](https://en.wikipedia.org/wiki/Flash_of_unstyled_content) related to the theme switches at runtime on statics, you have to build each page with a separate theme. This gets compounded by multilang support, so each language will also have to have a page per theme. And statically generated dynamic routes compound it even farther.
The basic solution I was thinking about is storing the theme keyword in a client-accessible cookie, read it somewhere in the middleware/server callback and return a different html  depending on its value. Not sure what can be done with statically generated dynamic routes though short of adding `theme` to params and then pretend it's a search query in the `Link` component.

**Reply by BierDav** [reply]
Not sure about your solution and how to implement it, but yes i think this should be a thing to consider. Also statistically generate pages in different languages. So that each CDN caches the language it should serve preferably

**Reply by GabenGar** [reply]
I was thinking about it and I've come up to the solution.
This approach makes these assumptions:
- the theme value is stored in a client-accessible cookie
- the theme is realized through CSS custom properties
- the theme value is also stored on the `<html>` element in a let's say `data-theme` attribute
- the values of these properties depend on the attribute value in `<html>` through the C in CSS.

To set the theme we need an inlined async script which reads the cookie value and sets it on `<html>` `data-theme` attribute.
NextJS has a [script element](https://nextjs.org/docs/basic-features/script), but its API requires to insert a string literal for our usecase. Also no modules within this script, so you'd have to deal with horrendous API of the parsing the entire cookie as a string.
But I think this snippet will suffice:
```js
(() => {
  const theme = document.cookie
    .split("; ")
    .find((row) => row.startsWith("theme="));
  const value = theme ? theme.split("=")[1] : "light";
  document.documentElement.dataset.theme = value;
})();
```
Due to being an async it will run before the full DOM gets parsed and most likely `<html>` is the only element available at the time of its call, hence why we store the theme in `<html>` instead of `<body>`. It's not a "true" SSR, but it's very likely to resolve way before the page gets rendered without flashes and works on static pages too with no overhead on the server.

The big con of this approach is if your theme system isn't written in CSS Custom Properties. Then it potentially requires a lot of work for rewriting the stylesheets and becomes impossible to pull off if you have a limited access to them.

**Reply by BierDav** [reply]
If the flash bang was my only concern I would have just created a static css that set the background color before rendering and hydration. But i want to improve the performance of my site with that.

**Reply by GabenGar** [reply]
How is the theme is going to affect performance in the first place?

**Reply by BierDav** [reply]
Startup time, Fetch duration, .... It depends on the solution on what, but without SSG somewhere the website geht's slow

**Comment by henrikvtcodes** [comment]
Have you considering using `prefers-color-scheme` in your CSS? I can't say that I know a lot about emotion, but here's my idea: Assuming that you have a color palette for your light and dark settings, you could make use of [reusable media queries](https://emotion.sh/docs/media-queries#reusable-media-queries) in emotion to grab the `prefers-color-scheme` property. Your components can then use these media queries to automatically switch themes. Unless you've already ruled this out, I think it's the best option because it's relatively simple and not dependent on javascript.

**Reply by BierDav** [reply]
Yes I have, but I want to control dark and light theme on my own. This makes testing easier.

**Comment by BierDav** [comment]
Can we expect a feature to support this or should we use React for simplicity

**Comment by GabenGar** [comment]
I've made a quick hack, you can check it out at https://gabengar.vercel.app and click `Light`/`Dark` button.
The workings are like this:
- the colors are defined as CSS Custom Properties
- their values depend on the [value of `data-theme` attribute of `<html>` element](https://github.com/GabenGar/next-js-site/tree/master/src/styles/variables/theme)
- the theme is [changed through `get()` and `set()` functions on client](https://github.com/GabenGar/next-js-site/blob/master/src/lib/theme/lib.ts), both of which interact with cookie value
- the inline script at `__app` page [reads the cookie value and sets it on `data-theme` attribute](https://github.com/GabenGar/next-js-site/blob/7b5da32d1b0adae85c72159e5653378f3f91344d/src/pages/_app.tsx#L33) (be sure to set it as `dangerouslySetInnerHTML` prop, inlining a string literal as `children` will apply a bunch of escaping to it)
- the theme is switched [in the nav bar by invoking `set()` function from above](https://github.com/GabenGar/next-js-site/blob/7b5da32d1b0adae85c72159e5653378f3f91344d/src/components/layout/base.tsx#L101)

You can click through any link at the top nav (all these pages are static), reload the page afterwards and see the theme being persisted and the page being rendered without a flash with no server logic for it.

**Reply by BierDav** [reply]
Thanks for your effort, is there a chance of getting the source code. Nevertheless can we expect this as a feature or not. We are not in the position to make a Pull request, because we don't have the resources, knowledge of the project and the perfect idea how to achieve this

**Reply by GabenGar** [reply]
The links within bullets are all source code, the hard part is to define your theme variables or port the existing theme framework to CSS Custom Properties.

**Reply by dickyindra** [reply]
Thanks for the quick hack!
I have implemented your steps, it works when moving pages without flash.
But I have a problem, when I reload, there is a flash from light mode to dark mode.

I saw your website https://gabengar.vercel.app/ , no flash on page reload, it's amazing.

**Reply by GabenGar** [reply]
Are you reloading in the dev or prod? In dev mode all pages are generated at request time and therefore can incur some delay on the theme switch script.

**Reply by mertafor** [reply]
Superb solution man! The problem with useEffect is it takes a lot of time to be initialized. Injecting JS into body directly in App resolved the problem since it's the first script runs. Thanks a lot!
